### PR TITLE
Relation widget

### LINKED ISSUE #61: Adjust relationship widget to provided mockups
@martomi starts with the search top bar (styles + provide callback when user types in)
@martomi implements API calls when user types in the search bar and populates data into state

My suggestion is for now to load top 50 relations + entities determined by server algorithm for given entity so that search could be quicker and pagination easier to implement. Search would act as a filter on React’s state in this case.
Later we can implement load on demand with search on server if we feel like the top 50 entities are not enough for the user to explore and there is enough time.

@lustoykov starts with adjusting and styling the table UI + Popups

I will expect data format like this:
```js
{ 
   entities: [ { title: '...' }, ... ],
   relations: [ { title: '...' }, ... ],
}
```

Next we'll implement the details view.

### PR DESCRIPTION
**Pull request by martomi**
Here's the first version of the relation widget as described in #56 . Resolves #61

### CODE REVIEW

**Comment by gyachdav** [comment]
Awesome! Is there a hosted demo somewhere?

**Comment by chaoran-chen** [comment]
[Here it is.](http://54.68.164.199:8080/relations-widget.html)

1200 LoC and all of the functionalities implemented - you were very hard-working on the holidays! :) I will review the code this evening. For now only a few comments to the popovers:
* The arrows at the bottom look a bit weird to me because they don't really point to anything. Maybe it's better to remove them.
* They should disappear when the user clicks outside of the popover.
* The source popover could maybe have a larger width (maybe 600-700px instead of 400?).
* Why didn't you want to reuse the [FeedbackComponent](https://github.com/MusicConnectionMachine/VisualizationG5/blob/develop/react/src/components/FeedbackComponent.jsx)? :(

Moreover, the download does not work and throws errors.

**Comment by gyachdav** [comment]
@chaoran-chen touvahuee this is the right demo? The component that got loaded in the URL you provided looks very similar to the original tabular component before we had the discussiion in #56.

**Comment by chaoran-chen** [comment]
unbenannt

Do you see this? I've just deployed it and I believe that it's the right implementation. @lustoykov, @martomi : can you confirm it?

**Comment by gyachdav** [comment]
Hmm.. yup I can see it but it is not really echoing the Spotify layout we discussed in #56. Is there a different presentation mode?

**Comment by lustoykov** [comment]
@gyachdav we interpreted your suggestion as an _enhancement_, which means visible only in full-screen mode. 
This is because there is not enough space to implement the suggested layout as a primary one. 
There is only as much height as about 300px on the website, which doesn't fit the concept of Spotify layout.

Therefore, in the previous discussion we have already mentioned (7 days ago #56) that we'll start with the most important feature first - the primary layout, and it was incredibly demanding.

I agree that the Spotify suggestion would be a killer feature and an incredible enhancement, but we're already past feature freeze and the tabular widget has bigger priority than a full-screen mode (@vviro can you confirm?), since it is the primary one. My suggestion is to concentrate on what we already have and polish it for production. You're very welcome to provide constructive bug reports similar to @chaoran-chen's, but anything beyond that would not fit in the timeline of the seminar.

**Comment by lustoykov** [comment]
@chaoran-chen thanks for the feedback, I'll take care to fix the popovers.

File: react/src/components/relations-widget/Application.jsx
Comment (lustoykov) [inline react/src/components/relations-widget/Application.jsx]: 💃

Comment (lustoykov) [reply react/src/components/relations-widget/Application.jsx]: 👯  *enjoying the fact that it was quick fix*

Comment (martomi) [reply react/src/components/relations-widget/Application.jsx]: As mentioned in the commit message, only exporting the first source from the list in this global export. If there's interest in exporting all sources for a given relation, we might add that download button inside the sources popup to export all of them separately.

Comment (lustoykov) [reply react/src/components/relations-widget/Application.jsx]: :D I think we have enough other stuff to do, I'm fine with that. (I was enjoying the quick fix **for real**)

**Comment by gyachdav** [comment]
@lustoykov thanks for the update. Can you list out all the tasks that are left to do for this widget? Would you be able to debug outstanding items before the hackathon?

**Comment by lustoykov** [comment]
@gyachdav we need to push 4 more Todo's and I think we'll complete them and merge just in time for the hackathon.

TODO: Integrate API with real world data (@chaoran-chen)
TODO: The arrows at the bottom look a bit weird to because they don't really point to anything. (@lustoykov)
TODO: Popover should disappear when the user clicks outside of the popover. (@lustoykov)
TODO: The source popover could maybe have a larger width (maybe 600-700px instead of 400?) (@lustoykov)

Also, Code Review: @chaoran-chen @Binabiko

**Comment by gyachdav** [comment]
TODO: Twitter share should have the text for the relation being shared and a link back to the referring page.

**Comment by martomi** [comment]
@gyachdav Actually, I did add the relation text in the twitter sharing, but it's not working for some reason. Will investigate. What other social sharing links do we want? Facebook, any others? Some more TODOs I can think of right now and will start working on:
1) The popovers can't be opened from the details view
2) There are some inconsistencies with mouse hover behaviour in main and details view
3) It's not possible to click on a relation / entity from a detail view

(By detail view, I refer to the view shown after clicking on a relation / entity from the main screen)

**Comment by chaoran-chen** [comment]
Some more suggestions:
* Use the Twitter image instead of writing "Twitter"
* Display the fixed entity bold.
* In the fullscreen mode, the complete height should be used.

File: react/src/components/relations-widget/Application.jsx
Comment (chaoran-chen) [inline react/src/components/relations-widget/Application.jsx]: Is this `entity` ever set and used?

**Comment by lustoykov** [comment]
OK I fixed the popovers, but couldn't fix the outside clicks, because the PopOvers are absolutely positioned and I can't properly catch the outside clicks. even if I click inside the popover the outside click event gets triggered. So you'd have to manually click close

**Comment by chaoran-chen** [comment]
@lustoykov, it is probably not the best way but you could catch all click events and check if the mouse position is outside of the popover using [getBoundingClientRect](https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect).

**Comment by lustoykov** [comment]
I've tried with [node.contains(event.target)](https://developer.mozilla.org/en/docs/Web/API/Node/contains), then with https://github.com/kentor/react-click-outside, there are two other libraries and your approach, but I lost my motivation, will try tomorrow on the hackathon.

**Comment by martomi** [comment]
All outstanding suggestions above should be implemented now.

1. The popovers can be opened from the details view
2. All view inconsistencies are fixed through a huge refactor
3. It's now possible to click on a relation / entity from a detail view
4. Sharing works from both main view and details views and includes the relation text as well as url back to the website (done for Twitter, Facebook and Google)

Here's what's left to do:

1. Connect the widget with the real API
2. Informs the user if something went wrong with loading the data or there's simply no data to show

**Comment by lustoykov** [comment]
Informing is done, only API left.

**Comment by martomi** [comment]
I think we're ready to merge that now?

File: react/src/RemoteService.js
Comment (chaoran-chen) [inline react/src/RemoteService.js:23]: Are you sure that this is correct and will not lead to wrong rejects? For example, when the readyState changes from OPENED to HEADERS_RECEIVED? See: https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/readyState

**Comment by martomi** [comment]
I just noticed at least 2 things that are broken after the merge. The header styling is a bit off and the entity is not shown on the top left.

**Comment by lustoykov** [comment]
Hi, thanks for the feedback, I pushed another version. Did it address your concerns?

File: react/src/RemoteService.js
Comment (lustoykov) [reply react/src/RemoteService.js:23]: `http.readyState === 4` should stay, I corrected it. Look again.

**Comment by chaoran-chen** [comment]
I will take a last look this night and approve if it all works. Is it okay?

**Comment by lustoykov** [comment]
I would love to merge this today, yes. Please write if there is something, I'll correct it asap.

**Review by chaoran-chen — APPROVED** [review]

**Comment by lustoykov** [comment]
🎉 🎉 🎉

**Pull request merged**
